gdb/rust-lang.c

```diff
@@ -60,7 +70,80 @@ rust_crate_for_block (const struct block *block)
   return xstrndup (scope, len);
 }
 
-
+/* Information about the discriminant/variant of an enum */
+
+struct disr_info
+{
+  char* name; // needs to be freed
+  int field_no; // Field number in union. negative if error
+};
+
+/* Utility function to get discriminant info for a given value.  */
+
+static struct disr_info
+rust_get_disr_info (struct type *type, const gdb_byte *valaddr,
+                    int embedded_offset, CORE_ADDR address,
+                    const struct value *val) {
+
+  int i;
+  struct disr_info ret;
+  struct type *disr_type;
+  struct ui_file * temp_file;
+
+  struct value_print_options opts;
+
+  get_no_prettyformat_print_options (&opts);
+
+  ret.field_no = -1;
+
+  if (TYPE_NFIELDS (type) == 0) {
+    /* void enums should never exist as values */
+    return ret;
+  }
+
+  disr_type = TYPE_FIELD_TYPE (type, 0);
+
+  if (TYPE_NFIELDS (disr_type) == 0) {
+    /* This is a bounds check and should never
+       be hit unless Rust has changed its debuginfo
+       format */
+    return ret;
+  }
+
+  if (strcmp (TYPE_FIELD_NAME (disr_type, 0), "RUST$ENUM$DISR") != 0) {
+    /* debuginfo format has changed */
+    return ret;
+  }
+
+  temp_file = mem_fileopen ();
+
```

Comment:
This is a spot that needs a cleanup.  `c_val_print` can throw (nearly anything in gdb can throw, it's usually the default); and if that happens then `temp_file` would leak.  See `make_cleanup_ui_file_delete`